# Predicting the credit risk associated with loans

In [3]:
import pandas as pd
loans_data = pd.read_csv('accepted_2007_to_2018Q4.csv.gz', compression='gzip', error_bad_lines=False,low_memory=False)
loans_data.shape

(2260701, 151)

### Check for missing values
Let's express the missing values per column as a percentage of the total length for easier interpretation.

In [7]:
missing_values = loans_data.isnull().sum()
missing_values = (missing_values*100)/loans_data.shape[0]
missing_values.sort_values(ascending=False)

member_id                                     100.000000
orig_projected_additional_accrued_interest     99.617331
hardship_loan_status                           99.517097
hardship_dpd                                   99.517097
hardship_reason                                99.517097
hardship_status                                99.517097
deferral_term                                  99.517097
hardship_amount                                99.517097
hardship_start_date                            99.517097
hardship_end_date                              99.517097
payment_plan_start_date                        99.517097
hardship_length                                99.517097
hardship_type                                  99.517097
hardship_payoff_balance_amount                 99.517097
hardship_last_payment_amount                   99.517097
settlement_percentage                          98.485160
debt_settlement_flag_date                      98.485160
settlement_status              

In [14]:
not_missing = missing_values[missing_values <= 10]
not_missing.sort_values(ascending=False)

emp_title                     7.387178
num_tl_120dpd_2m              6.798334
emp_length                    6.499754
mo_sin_old_il_acct            6.153136
bc_util                       3.366389
percent_bc_gt_75              3.335779
bc_open_to_buy                3.316140
mths_since_recent_bc          3.248771
pct_tl_nvr_dlq                3.116909
avg_cur_bal                   3.113149
mo_sin_rcnt_rev_tl_op         3.110097
mo_sin_old_rev_tl_op          3.110097
num_rev_accts                 3.110097
num_bc_tl                     3.110053
mo_sin_rcnt_tl                3.110053
tot_coll_amt                  3.110053
num_accts_ever_120_pd         3.110053
num_actv_bc_tl                3.110053
num_actv_rev_tl               3.110053
num_il_tl                     3.110053
tot_cur_bal                   3.110053
num_tl_90g_dpd_24m            3.110053
total_il_high_credit_limit    3.110053
tot_hi_cred_lim               3.110053
num_tl_op_past_12m            3.110053
total_rev_hi_lim         

~61% of the columns (92 out of 151)  in the dataset contain less than 10% total missing values. We can purge the dataset of the columns that contain more than 10% missing values as it inhibits our capability to make good predictions.

In [16]:
loans_data = loans_data[not_missing.index]
loans_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2260701 entries, 0 to 2260700
Data columns (total 92 columns):
id                            object
loan_amnt                     float64
funded_amnt                   float64
funded_amnt_inv               float64
term                          object
int_rate                      float64
installment                   float64
grade                         object
sub_grade                     object
emp_title                     object
emp_length                    object
home_ownership                object
annual_inc                    float64
verification_status           object
issue_d                       object
loan_status                   object
pymnt_plan                    object
url                           object
purpose                       object
title                         object
zip_code                      object
addr_state                    object
dti                           float64
delinq_2yrs                  